examples/transactions.c

```diff
@@ -602,67 +606,70 @@ int main (int argc, char **argv) {
+                if (msgcnt > 0 &&
+                    (msgcnt > 100 || last_commit + 5 <= time(NULL))) {
+                        printf("msgcnt %d, elapsed %d\n", msgcnt,
+                               (int)(time(NULL) - last_commit));
+                        commit_transaction_and_start_new(consumer, producer);
+                        msgcnt = 0;
+                        last_commit = time(NULL);
+                }
 
                 /* Wait for new mesages or error events */
                 msg = rd_kafka_consumer_poll(consumer, 1000/*1 second*/);
                 if (!msg)
-                        continue;
+                        continue; /* Poll timeout */
 
                 if (msg->err) {
                         /* Client errors are typically just informational
                          * since the client will automatically try to recover
                          * from all types of errors.
                          * It is thus sufficient for the application to log and
-                         * continue operating when an error is received. */
+                         * continue operating when a consumer error is
+                         * encountered. */
                         fprintf(stderr, "WARNING: Consumer error: %s\n",
                                 rd_kafka_message_errstr(msg));
                         rd_kafka_message_destroy(msg);
                         continue;
                 }
 
-                /* Find output producer for this input partition */
-                rktpar = rd_kafka_topic_partition_list_find(
-                        assigned_partitions,
-                        rd_kafka_topic_name(msg->rkt), msg->partition);
-                if (!rktpar)
-                        fatal("BUG: No output producer for assigned "
-                              "partition %s [%d]",
-                              rd_kafka_topic_name(msg->rkt),
-                              (int)msg->partition);
-
-                /* Get state struct for this partition */
-                state = (struct state *)rktpar->opaque;
-
                 /* Process message */
-                process_message(state, msg);
+                process_message(consumer, producer, output_topic, msg);
 
                 rd_kafka_message_destroy(msg);
 
-                /* Commit transaction every 100 messages or 5 seconds */
-                if (++state->msgcnt > 100 ||
-                    state->last_commit < time(NULL) + 5) {
-                        commit_transaction_and_start_new(state);
-                        state->msgcnt = 0;
-                        state->last_commit = time(NULL);
-                }
+                msgcnt++;
         }
 
         fprintf(stdout, "Closing consumer\n");
         rd_kafka_consumer_close(consumer);
-
         rd_kafka_destroy(consumer);
 
+        fprintf(stdout, "Closing producer\n");
+        rd_kafka_destroy(producer);
+
         return 0;
 }
```

Comment:
logic seem fine

the main thing I feel is i want more clarity on error handling / timeouts. i think the way to do it is try and compress a set of general ideas into some blurb somewhere, so as to create a mental model for thinking about this and have commentary in specific places linking back to that.